### PR TITLE
[MIR] Implement coercions and improve fat pointer support

### PR DESCRIPTION
**Pull request by arielb1**
I am also planning to implement casts, but I will probably do it in a separate PR.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_trans/trans/mir/rvalue.rs [hunk 1]
```diff
@@ -45,8 +45,36 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+                // this to be eliminated by MIR translation, but
+                // `CoerceUnsized` can be passed by a where-clause,
+                // so the (generic) MIR may not be able to expand it.
+                let operand = self.trans_operand(bcx, operand);
+                match operand.val {
+                    OperandValue::FatPtr(..) => unreachable!(),
+                    OperandValue::Imm(llval) => {
+                        // ugly alloca.
+                        debug!("trans_rvalue: creating ugly alloca");
+                        let lltemp = base::alloc_ty(bcx, operand.ty, "__unsize_temp");
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/rvalue.rs]: Question out of curiosity: Is there any reason this should be alloca and cannot be regular LLVM temporary variable?

Comment (nikomatsakis) [inline src/librustc_trans/trans/mir/rvalue.rs]: @nagisa I think the reason is that `base::coerce_unsized_into` expects its argument to be by-reference.

**Comment by nikomatsakis** [comment]
OK, so, this code looks good. r+ from me, but I left a few nits and things. Nothing major. The only thing I'd say let's definitely do is change the variant from `Imm` to `Immediate`. And of course if you feel the desire to write some more docs, that can't hurt too :), but I wouldn't block the PR on it.

One thing that might be nice is a few more tests. For example, I didn't see many tests where the MIR code actually does the unsizing, e.g.:

```rust
#[rustc_mir]
fn foo(x: &[i32; 3]) -> &[i32] {
    let y: &[i32] = x;
    y
}
```

Was that around, or did I miss it?

**Comment by arielb1** [comment]
@nikomatsakis 

There is the `fn_coercions` test in `mir_coercions`, but I will add one for arrays too.

**Comment by arielb1** [comment]
<updated>

**Comment by arielb1** [comment]
⟨rebased⟩

**Pull request merged**